src/main/native/org_cryptomator_windows_autostart_WinShortcutCreation_Native.cpp

```diff
@@ -0,0 +1,117 @@
+#ifndef UNICODE
+    #define UNICODE
+#endif
+
+#include <jni.h>
+#include <windows.h>
+#include <winnls.h>
+#include <shobjidl.h>
+#include <objbase.h>
+#include <objidl.h>
+#include <shlguid.h>
+#include <wchar.h>
+#include "org_cryptomator_windows_autostart_WinShortcutCreation_Native.h"
+
+LPCWSTR get_nullterminate_and_release(JNIEnv * env,  jstring name);
+HRESULT CreateLink(LPCWSTR lpszPathObj, LPCWSTR lpszPathLink, LPCWSTR lpszDesc);
+
+JNIEXPORT jint JNICALL Java_org_cryptomator_windows_autostart_WinShortcutCreation_00024Native_createShortcut
+  (JNIEnv * env, jobject thisObj, jstring target, jstring storage_path, jstring description) {
+
+    //make the utf16 char arrays null terminated
+    LPCWSTR link_target = get_nullterminate_and_release(env,target);
+    LPCWSTR link_location = get_nullterminate_and_release(env,storage_path);
+    LPCWSTR link_description = get_nullterminate_and_release(env,description);
+
+    //initialize
+    int hResult = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED);
+    if( hResult != S_OK && hResult != S_FALSE && hResult != RPC_E_CHANGED_MODE) {
+        //error of the type E_INVALIDARG, E_OUTOFMEMORY or E_UNEXPECTED
```

Comment:
Should unfortunately leak memory here upon failure (`link_target`, `link_location` and `link_description`).

Reply:
Because `CoUninitialize()` is not called?

Reply:
On another note, this is outdated (-;

Reply:
But same argument applies to the current impl:

https://github.com/cryptomator/integrations-win/blob/3da0465e609a06995be42bfa8f5e5f6c422032e8/src/main/native/org_cryptomator_windows_autostart_WinShellLinks_Native.cpp#L26-L28

The cleanup phase is not run when returning here.

Reply:
According to the Microsoft, [`CoUninitialize()`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize#remarks) needs only be called, if the method "successfully" returns (meaning it is one of the checked return values).

But yeah, the array releases could be still done.